[quant][fx] Relax the constraint for input of custom module nodes

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #74510

Summary:
Previously we require the dequantize before custom module to have one user, this is because we are removing the dequantize node
before custom module while we transform an observed custom module to a quantized custom module, but actually we don't need to remove it,
we can just change the input of custom module with quantize node instead. If the dequantize node only has one user, it will be removed. We are removing the node directly instead of relying on dead code elimination to improve the debugability of the system, since people will see the final graph after the conversion, instead of introducing the dce as a step that changes the graph. We will remove the dependence on DCE in the future as well.

Test Plan:
python test/test_quantization.py TestQuantizeFx.test_custom_module_class_input_has_multiple_users

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D35034626](https://our.internmc.facebook.com/intern/diff/D35034626)

jerryzh168 committed:
[quant][fx] Relax the constraint for input of custom module nodes

Summary:
Previously we require the dequantize before custom module to have one user, this is because we are removing the dequantize node
before custom module while we transform an observed custom module to a quantized custom module, but actually we don't need to remove it,
we can just change the input of custom module with quantize node instead. If the dequantize node only has one user, it will be removed
by the dead code elimination pass that was added recently.

Test Plan:
python test/test_quantization.py TestQuantizeFx.test_custom_module_class_input_has_multiple_users

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35034626).

vkuzo commented on torch/ao/quantization/fx/convert.py:
```diff
@@ -533,11 +533,15 @@ def convert_custom_module(
+            #              \ - dequantize
+            #                  (may be dangling depending on how many users it has before)
+            # if the custom module node is the only user of dequantize node, then the
+            # dequantize node will be remove in the dead code elimination pass
```
What else are we using DCE for?  IMO using it like this makes things harder to understand, it would seem cleaner to me to always have local graph rewrites have a "perfect" end state without relying on DCE to clean up the whole graph at the end.

jerryzh168 replied:
this is the only thing right now I think. I feel relying on DCE would help us to reduce the complexity of our passes, otherwise we'd need to duplicate all the node that might appear in patterns, like dequantize, quantize_per_tensor_dynamic, quantize_per_tensor etc.

vkuzo commented on torch/ao/quantization/fx/convert.py:
```diff
@@ -533,11 +533,15 @@ def convert_custom_module(
+            # Before: quantize - dequantize - custom - module
+            # After: quantize - custom - module
+            #              \ - dequantize
+            #                  (may be dangling depending on how many users it has before)
```
can we just do the check here to clean it up instead of relying on DCE?  This way the intermediate state will be easier to understand, for any developers that need to work on the intermediate state (to fix a bug, to develop a new feature, etc)

jerryzh168 replied:
I see, yeah I think we can do that as well

jerryzh168 committed:
Update on "[quant][fx] Relax the constraint for input of custom module nodes"


Summary:
Previously we require the dequantize before custom module to have one user, this is because we are removing the dequantize node
before custom module while we transform an observed custom module to a quantized custom module, but actually we don't need to remove it,
we can just change the input of custom module with quantize node instead. If the dequantize node only has one user, it will be removed
by the dead code elimination pass that was added recently.

Test Plan:
python test/test_quantization.py TestQuantizeFx.test_custom_module_class_input_has_multiple_users

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D35034626](https://our.internmc.facebook.com/intern/diff/D35034626)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35034626).

jerryzh168 commented:
thanks for the suggestion @vkuzo, I think it makes sense, also added an item in our BE task list to remove DCE in our code base

andrewor14 reviewed (approved):
Does the latest code not rely on DCE? If so I think the PR description is out of date now.

jerryzh168 commented:
> Does the latest code not rely on DCE? If so I think the PR description is out of date now.

yeah, we remove the dequantize op directly after that now, will update the PR description
